src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+                trampoline, overload, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets the parameter name, if the expression's grandparent is a variable declarator then it just gets the
+        /// local declarations name. Otherwise, it generates a name based on the context of the expression.
+        /// </summary>
+        private async Task<string> GetNewParameterNameAsync(Document document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (ShouldRemoveVariableDeclaratorContainingExpression(document, expression, out var varDeclName, out _))
+            {
+                return varDeclName;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(semanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Determines if the expression's grandparent is a variable declarator and if so,
+        /// returns the name
+        /// </summary>
+        protected bool ShouldRemoveVariableDeclaratorContainingExpression(
+            Document document, TExpressionSyntax expression, [NotNullWhen(true)] out string? varDeclName, [NotNullWhen(true)] out SyntaxNode? localDeclaration)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var expressionDecl = expression?.Parent?.Parent;
+
+            localDeclaration = null;
+
+            if (!syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            localDeclaration = GetLocalDeclarationFromDeclarator(expressionDecl);
+            if (localDeclaration is null)
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            // TODO: handle in the future
+            if (syntaxFacts.GetVariablesOfLocalDeclarationStatement(localDeclaration).Count > 1)
+            {
+                varDeclName = null;
+                localDeclaration = null;
+                return false;
+            }
+
+            varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+            return true;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        protected static async Task<ImmutableDictionary<Document, List<SyntaxNode>>> FindCallSitesAsync(
+            Document document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSitesBuilder = ImmutableDictionary.CreateBuilder<Document, List<SyntaxNode>>();
```

Comment:
```suggestion
            var methodCallSitesBuilder = new Dictionary<Document, List<SyntaxNode>>();
```

A few reasons for this:

1. you're not storing this anywhere, so it's ok that's mutable
2. it's effectively temp state.  so building it up just to do this full copy isn't really necessary.
3. the dictionary already points at mutable values (List<SyntaxNode>) so it's weirdly inconsistent.  Makign the values immutable is overkill, so just embrace mutable here :)

Reply:
embracing the mutable now